geth/api/backend.go

```diff
@@ -77,120 +78,97 @@ func (m *StatusBackend) IsNodeRunning() bool {
-	<-nodeStarted
-
+func (m *StatusBackend) onNodeStart() {
 	// tx queue manager should be started after node is started, it depends
 	// on rpc client being created
 	m.txQueueManager.Start()
-
 	if err := m.registerHandlers(); err != nil {
 		log.Error("Handler registration failed", "err", err)
 	}
-
 	if err := m.accountManager.ReSelectAccount(); err != nil {
 		log.Error("Reselect account failed", "err", err)
 	}
 	log.Info("Account reselected")
-
-	close(backendReady)
-	signal.Send(signal.Envelope{
-		Type:  signal.EventNodeReady,
-		Event: struct{}{},
-	})
 }
 
 // StopNode stop Status node. Stopped node cannot be resumed.
-func (m *StatusBackend) StopNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
+func (m *StatusBackend) StopNode() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.stopNode()
+}
 
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
+func (m *StatusBackend) stopNode() error {
+	if !m.IsNodeRunning() {
+		return node.ErrNoRunningNode
 	}
-	<-m.nodeReady
-
 	m.txQueueManager.Stop()
 	m.jailManager.Stop()
-
-	nodeStopped, err := m.nodeManager.StopNode()
-	if err != nil {
-		return nil, err
-	}
-
-	backendStopped := make(chan struct{}, 1)
-	go func() {
-		<-nodeStopped
-		m.Lock()
-		m.nodeReady = nil
-		m.Unlock()
-		close(backendStopped)
-	}()
-
-	return backendStopped, nil
+	return m.nodeManager.StopNode()
 }
 
 // RestartNode restart running Status node, fails if node is not running
-func (m *StatusBackend) RestartNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
-	}
-	<-m.nodeReady
-
-	nodeRestarted, err := m.nodeManager.RestartNode()
-	if err != nil {
-		return nil, err
-	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeRestarted, m.nodeReady) // waits on nodeRestarted, writes to backendReady
-
-	return m.nodeReady, err
+func (m *StatusBackend) RestartNode() error {
+	return m.nodeManager.RestartNode()
 }
 
 // ResetChainData remove chain data from data directory.
 // Node is stopped, and new node is started, with clean data directory.
-func (m *StatusBackend) ResetChainData() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
-	}
-	<-m.nodeReady
-
-	nodeReset, err := m.nodeManager.ResetChainData()
+func (m *StatusBackend) ResetChainData() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	config, err := m.nodeManager.NodeConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeReset, m.nodeReady) // waits on nodeReset, writes to backendReady
-
-	return m.nodeReady, err
+	newcfg := *config
+	if err := m.stopNode(); err != nil {
+		return err
+	}
+	if err := m.ResetChainData(); err != nil {
+		return err
+	}
+	signal.Send(signal.Envelope{Type: signal.EventChainDataRemoved})
+	if err := m.startNode(&newcfg); err != nil {
```

Comment:
since you are calling the manager's methods here, we won't receive
`signal.Envelope{Type: signal.EventNodeStarted}`
`signal.Send(signal.Envelope{Type: signal.EventNodeReady}`
and all the accounts trickery here.

if we don't want that, it would be nice to add a comment, why

Reply:
oh thanks, we definitely want it. in my previous changes i got rid from RestartNode on a manager, and then forgot why

Reply:
@adambabik that was an issue with restart node, I need to move it to StatusBackend because otherwise, signals will have to be declared in node manager